File: 14_FMDBDemo/FMDBDemo_oc/FMDBDemo_oc/ViewController.m
//
//  ViewController.m
//  FMDBDemo_oc
//
//  Created by sunke on 2020/9/19.
//  Copyright © 2020 KentSun. All rights reserved.
//

#import "ViewController.h"
#import "HttpTool.h"
#import "User.h"
#import "HomeCell.h"

@interface ViewController ()
@property (nonatomic, strong) NSMutableArray *datasource;

@end

@implementation ViewController


#pragma mark - 懒加载
- (NSMutableArray *)datasource {
    if (_datasource == nil) {
        _datasource = [[NSMutableArray alloc] init];
    }
    return _datasource;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // 从缓存获取数据
    NSArray *array = [User fetchDatasource];
    if (array.count > 0) {
        
        [self.datasource addObjectsFromArray:array];
        NSLog(@"从缓存加载数据");
        return;
    }
    
    // 网络加载数据
    NSLog(@"从网络加载数据");
    [HttpTool loadRequestWithUrlString:@"http://api.budejie.com/api/api_open.php?a=list&c=data&type=29" resultData:^(NSDictionary *resultData) {
        NSArray *dictArray = resultData[@"list"];
        for (NSDictionary *dict in dictArray) {
            User *user = [User dict:dict];
            [self.datasource addObject:user];
            [user insert];
        }
        [self.tableView reloadData];
    }];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.datasource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *ID = @"HomeCell";
    HomeCell *cell = [tableView dequeueReusableCellWithIdentifier:ID forIndexPath:indexPath];
    cell.user = self.datasource[indexPath.row];
    return  cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
     return [[self.datasource[indexPath.row] detail].text boundingRectWithSize:CGSizeMake(self.view.bounds.size.width - 20, MAXFLOAT) options:NSStringDrawingUsesLineFragmentOrigin attributes:@{NSFontAttributeName : [UIFont systemFontOfSize:17]} context:nil].size.height + 87;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        User *user = self.datasource[indexPath.row];
        [self.datasource removeObject:user];
        [tableView reloadData];
    }
}


@end

File: 14_FMDBDemo/FMDBDemo_swift_error/FMDBDemo_swift/Controller/ViewController.swift
//
//  ViewController.swift
//  FMDBDemo_swift
//
//  Created by sunke on 2020/9/19.
//  Copyright © 2020 KentSun. All rights reserved.
//

import UIKit

class ViewController: UITableViewController {
    
    var dataSource = [User]() {
        didSet {
            self.cellHeights = getCellHeight()
            tableView.reloadData()
        }
    }
    
    lazy var cellHeights: [CGFloat] = []
    
    func getCellHeight() -> [CGFloat] {
        return self.dataSource.map{
            let detailText = $0.detail?.text ?? ""
            let textHeight = detailText.boundingRect(with: CGSize(width: self.view.bounds.width - 20, height: CGFloat(MAXFLOAT)), options: .usesLineFragmentOrigin, attributes: [NSAttributedString.Key.font : UIFont.systemFont(ofSize: 17)], context: nil).size.height
            return detailText.isEmpty ? 0 : textHeight + 87
        }
        
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        // 从数据库中获取数据
        User.dataSource {
            self.dataSource = $0
        }
        
        if dataSource.count != 0 {
            print("从缓存中获取")
            return
        }
        
        print("从网络获取数据")
        // 发起网络请求
        HttpTool.loadRequest(urlString: "http://api.budejie.com/api/api_open.php?a=list&c=data&type=29", method: .get) { (result, error) in
            if error != nil {
                return
            }
            
            guard let dictArray = result["list"] as? [[String : Any]] else {
                return
            }
            
            self.dataSource = dictArray.map{
                let user = User(dict: $0)
                user.insert()
                return user
            }
        }
    }
}


// MARK: - UITableView datasource
extension ViewController {
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataSource.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "HomeCell", for: indexPath) as! HomeCell
        cell.user = dataSource[indexPath.row]
        return cell
    }
}

// MARK: - UITableView delegate
extension ViewController {
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return cellHeights[indexPath.row]
    }
    
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            let user = dataSource[indexPath.row]
            user.detail?.delete()
            
            dataSource.remove(at: indexPath.row)
            tableView.reloadData()
            
        }
        
    }
    
    override func tableView(_ tableView: UITableView, titleForDeleteConfirmationButtonForRowAt indexPath: IndexPath) -> String? {
        return "删除"
    }
}
